bpf/bpf_host.c

```diff
@@ -968,6 +976,21 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 __section("from-netdev")
 int from_netdev(struct __ctx_buff *ctx)
 {
+	__u32 __maybe_unused vlan_id;
+
+	/* Filter allowed vlan id's and pass them back to kernel.
+	 */
+	if (ctx->vlan_present) {
```

Comment:
Have you checked whether this whole if-statement gets optimized away by the compiler if no vlans are specified? If not, then we would need to wrap this into #ifdef.

Reply:
vlan_allow function is always inline and it defenitely will be optimized, cause the code will look like:

```
if (false) {
  return CTX_ACT_OK;
} else {
  return send_drop_notifiy(...);
}
```

But don't know how to check this...

Reply:
You can use `llvm-objdump -S`. You can find the relevant object file it in `/var/run/cilium/state`. If it's not optimized, maybe the verifier will remove the branch. For this `bpftool prog dump` is useful.

Reply:
Thanks. I will check this and will get back to you. Meanwhile checking all other changes for 'problems' will be appreciated :)

Reply:
The rest LGTM. I'd appreciate if you would come up with a testing strategy too.

Reply:
I'm testing locally on VM's, but my install is specific - gentoo linux... I think only testing on VM's will help here, but to write such test will need to learn a lot about cilium's testing infastructure...

Reply:
I've checked - llvm-objdump is showing that this function is always optimized. If statements are removed in case they are not needed (always false or always true).

Reply:
Marking this conversation as resolved cause of this:

```
       4:	54 01 00 00 ff 0f 00 00	w1 &= 4095
       5:	16 01 1b 00 00 00 00 00	if w1 == 0 goto +27 <LBB6_6>
       6:	b4 01 00 00 02 00 00 00	w1 = 2
       7:	63 16 40 00 00 00 00 00	*(u32 *)(r6 + 64) = r1
       8:	b4 01 00 00 4a ff ff ff	w1 = -182
       9:	63 16 38 00 00 00 00 00	*(u32 *)(r6 + 56) = r1
      10:	b4 01 00 00 00 00 00 00	w1 = 0
```

We can see that there is no ACT_OK return here. Function is optimized.